src/c_api.cpp

```diff
@@ -364,9 +373,13 @@ class Booster {
  private:
   const Dataset* train_data_;
   std::unique_ptr<Boosting> boosting_;
-  std::unique_ptr<Predictor> single_row_predictor_;
-  PredictFunction single_row_predict_function_;
-  int64_t single_row_num_pred_in_one_row_;
+  std::unique_ptr<Predictor> single_row_predictor_[PREDICTOR_TYPES];
+  bool single_row_predictor_pred_early_stop[PREDICTOR_TYPES];
+  int single_row_predictor_pred_early_stop_freq[PREDICTOR_TYPES];
```

Comment:
just to confirm, will these arrays be freed correctly with the class on the heap once class is deleted?

Reply:
those are fixed size arrays and they'll be deallocated directly.

I'd suggest to use create a 

# new class containing (unique_ptr, bool, int) 
# implement operator== (const Config&) and operator=(const Config&)
# would make the assignment and comparison a bit more readable.